### PR TITLE
Add mock expectation support

### PR DESCRIPTION
**Pull request by ryanricard**
Initial implementation of adding expectation interface to all api methods. 

Example:

```js
const expectation = mockyeah
  .post('/foo', { text: 'bar' })
  .expect()
  .header('host', 'example.com')
  .params({
    id: '9999'
  })
  .body({
    foo: 'bar'
  })
  .once();

request
  .post('/foo?id=9999')
  .set('HOST', 'example.com')
  .send({ foo: 'bar' })
  .end(() => {
    expectation.verify();
    done();
  });
```

### CODE REVIEW

File: test/integration/RouteExpectationTest.js [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+    request
+      .post('/foo?id=9999')
+      .set('HOST', 'example.com')
+      .send({ foo: 'bar' })
+      .end(() => {
+        expectation.verify();
+        done();
+      });
+  });
+});
```
Comment (therynamo) [inline test/integration/RouteExpectationTest.js]: Wonderful Tests! This definitely helps in figuring out our API 👍

**Comment by mathume** [comment]
Hi! I think this is a good API.

Could you eventually add a request journal dump, too? The use case is that during development sometimes I want to inspect the request journal (diagnostics and so on). It's a feature similar to mockyeah-cli ls, but in the test development phase we normally only run mockyeah from within the tests and there's "no time" to invoke mockyeah-cli ls.

If I can be of any help, I'd be glad to lend a hand.

**Comment by therynamo** [comment]
@mathume, is what you want more of just an assertion style log?

example:
```
- testing route foo/bar
``` Or are you asking for a physical file that contains logs of which routes were hit?

File: app/lib/Expectation.js [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    exactly: function exactly(number) {
+      internal.assertions.push(() => {
+        assert(internal.called === number, `${internal.prefix} Expected route to be called ${number} times, but it was called ${internal.called} times`);
+      });
+      return this;
+    },
+    header: function header(name, value) {
+      internal.handlers.push((req) => {
+        const actualValue = req.get(name);
+        assert.equal(actualValue, value, `${internal.prefix} Expected header value ${name}:${value}, but it was ${actualValue}`);
```
Comment (ryanricard) [inline app/lib/Expectation.js]: @therynamo redundant getter fixed with 7ef1289ed1512354f7b2663b49e0a96cfcd98f62.

**Comment by therynamo** [comment]
@mathume, could you potentially clarify what you're asking for? 

Are you interested in more verbose logging when using this api, or something else?

File: app/lib/Expectation.js (hunk 2 above)
Comment (therynamo) [inline app/lib/Expectation.js]: Sounds good. 👍

**Comment by therynamo** [comment]
@mathume, please open an issue if you'd like to further discuss.

**Comment by ryanricard** [comment]
@mathume absolutely, your request for a request journal is making more sense to me now. I apologize, I didn't previously investigate the resources you linked to in https://github.com/ryanricard/mockyeah/issues/28. Let's converse over the issue to determine the design.

**Comment by mathume** [comment]
Oh, sorry, of course.

Normally, when I implement tests the following could happen:

I would expect something like:

TEST SETUP (e.g. beforeEach)
mockyeah.get(/somePath/, {json:someResponseText});

TEST IMPLEMENTATION (e.g. it('should give me a return value X when Y',...)
myapp.doSomeStuffThatWouldCauseACallTo(/somePath/);
expectThat(myapp.someObject.somePropertyHasBeenCorrectlySetByResponse).to.be.true;

TEST RUN => FAILS
TEST IMPLEMENTATION (e.g. it('should give me a return value X when Y',...)
myapp.doSomeStuffThatWouldCauseACallTo(/somePath/);
console.log(mockyeah.requests.dump());
expectThat(myapp.someObject.somePropertyHasBeenCorrectlySetByResponse).to.be.true;

ConsoleOutput:
{ '/api/somePath', requestBody: { property: "value" }, requestHeaders: { "Content-Type" : "application-json", "Accept" : "application-json" } }....

FIX TEST SETUP:
mockyeah.get(/api\/somePath/, {json:someResponseText});

OR

FIX API /api/somePath -> /somePath

**Pull request merged**